### PR TITLE
Add Falcon HF model

### PR DESCRIPTION
**Pull request by rmitsch**
## Description

Add Falcon HF model.
Notes: 
- Based on `refactor/backend-to-models`. Currently aimed towards that branch for easier comparison.
- Readme will be updated in `refactor/backend-to-models`, so the readme in change in this PR aren't final.

### Types of change

New feature.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran all tests in `tests` and `usage_examples/tests`, and all new and existing tests passed. This includes
  - all external tests (i. e. `pytest` ran with `--external`)
  - all tests requiring a GPU
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by rmitsch** [comment]
[Succeeding GPU tests](https://buildkite.com/explosion-ai/spacy-llm-gpu-tests/builds/92#0188d977-7df4-4dda-bcea-7b5217f1e58c).

**Review by svlandeg — COMMENTED** [review]
Looks good! Let's get the refactor done first and then finalize this one.

File: README.md [hunk 1]
```diff
@@ -897,6 +896,49 @@ Note that Hugging Face will download this model the first time you use it - you
...
+
+Supported models:
+
+- `"rw-1b"`
+- `"7b"`
+- `"7b-instruct"`
+- `"40b-instruct"`
+
+Note that Hugging Face will download this model the first time you use it - you can
+[define the cached directory](https://huggingface.co/docs/huggingface_hub/main/en/guides/manage-cache)
```
Comment (shadeMe) [inline README.md:939]:
```suggestion
[define the cache directory](https://huggingface.co/docs/huggingface_hub/main/en/guides/manage-cache)
```

File: spacy_llm/models/hf/falcon.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+class Falcon(HuggingFace):
+    def __init__(
+        self,
+        variant: str,
+        config_init: Optional[Dict[str, Any]],
+        config_run: Optional[Dict[str, Any]],
+    ):
+        self._tokenizer: Optional["transformers.AutoTokenizer"] = None
+        self._device: Optional[str] = None
+        super().__init__(
```
Comment (shadeMe) [inline spacy_llm/models/hf/falcon.py:19]: Is there a reason why this isn't first statement in the ctor?

File: spacy_llm/models/hf/falcon.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+def falcon_hf(
+    variant: Literal["rw-1b", "7b", "7b-instruct", "40b-instruct"],  # noqa: F722
+    config_init: Optional[Dict[str, Any]] = SimpleFrozenDict(),
+    config_run: Optional[Dict[str, Any]] = SimpleFrozenDict(),
+) -> Callable[[Iterable[str]], Iterable[str]]:
+    """Generates Falcon instance that can execute a set of prompts and return the raw responses.
+    variant (Literal): Name of the Falcon model variant. Has to be one of Falcon.get_model_variants().
+    config_init (Optional[Dict[str, Any]]): HF config for initializing the model.
+    config_run (Optional[Dict[str, Any]]): HF config for running the model.
+    RETURNS (Callable[[Iterable[str]], Iterable[str]]): Dolly instance that can execute a set of prompts and return
```
Comment (shadeMe) [inline spacy_llm/models/hf/falcon.py:75]:
```suggestion
    RETURNS (Callable[[Iterable[str]], Iterable[str]]): Falcon instance that can execute a set of prompts and return
```

**Review by shadeMe — COMMENTED** [review]
Some drive-by reviewing.

File: spacy_llm/models/hf/falcon.py (hunk 2 above)
Comment (rmitsch) [reply spacy_llm/models/hf/falcon.py:19]: The parent constructor calls the `init_model()` method, which in this subclass accesses `self._tokenizer` and `self._device`.

File: README.md [hunk 4]
```diff
@@ -1182,6 +1196,42 @@ Note that Hugging Face will download this model the first time you use it - you
 [define the cached directory](https://huggingface.co/docs/huggingface_hub/main/en/guides/manage-cache)
 by setting the environmental variable `HF_HOME`.
 
+#### spacy.Falcon.HF.v1
```
Comment (svlandeg) [inline README.md:1199]:
```suggestion
#### spacy.Falcon.v1
```

File: spacy_llm/models/integration/__init__.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+from .remote.langchain import model_langchain, query_langchain
+from .remote.minichain import model_minichain, query_minichain
+
+__all__ = [
+    "Remote",
+    "query_minichain",
+    "query_langchain",
+    "model_minichain",
+    "model_langchain",
+]
```
Comment (svlandeg) [inline spacy_llm/models/integration/__init__.py:11]: Do I remember correctly that we had removed this `integration` subdir? (merge conflict leftover?)

**Review by svlandeg — COMMENTED** [review]
Few small issues, though otherwise close to merge. 

Do still need to discuss how to sync with https://github.com/explosion/spaCy/pull/12782 ;-)

File: spacy_llm/models/integration/__init__.py (hunk 5 above)
Comment (rmitsch) [reply spacy_llm/models/integration/__init__.py:11]: Yes, seems to have been a merge slip-up.

Comment (rmitsch) [reply spacy_llm/models/integration/__init__.py]: Removed in https://github.com/explosion/spacy-llm/pull/179/commits/88b2c8f2aeb78242bf26304caefae918b9478c12.

**Pull request merged**